### ISSUE TITLE
Retyping master password & GUI

### ISSUE THREAD

**Issue by sumproxy** [conradkleinespel/rooster#2]
I like to have a long master password, however without GUI I have to retype it for every operation, which is really bothersome. Is there a solution to this?

**Comment by conradkleinespel** [comment]
Hi @sumproxy,

Thanks a lot for your question. It is a good one ! :+1: 

Unfortunately, **there is no solution for this at the moment**. Having to retype the password is what I initially aimed for, because it keeps security rather high. A **web UI is on the road map, but might take a while** to be ready, because Rooster has no full-time employees at the moment.

If you have any suggestions as to what features you'd like to have in the GUI, feel free to let us know :smiley: 

Best regards,
Conrad

**Comment by sumproxy** [comment]
I'm fine with basic functionality, just ship what you can as soon as you can. Looking forward to it :smile:

**Comment by conradkleinespel** [comment]
@sumproxy Glad to hear you find Rooster useful as is ! And thanks for your support. I will be working on this in the coming months. Sorry if this takes time, but I'll do my best :muscle: 

I'll leave this issue open in case someone wants to add something down the road. I've changed the title of the issue to include "GUI".

**Comment by conradkleinespel** [comment]
Hey @sumproxy,

How have you been ? Happy new Year !

This message is coming a bit late: I'm sorry, I didn't have much time to work on Rooster. However, building a GUI is one of the things that Rooster should get, [now that Rooster v2 is out](https://github.com/conradkleinespel/rooster/releases/tag/v2.0.0).

Did you have any ideas in mind ? Any things you would like to work on yourself ? I would be happy to help you get started :smiley: 

Take care,
Conrad

**Comment by sumproxy** [comment]
Hi Conrad,

Happy new Year and Merry Christmas! If you celebrate :)
I’ve been flattered by your invitation to take part in development, but I’m still a noob and do not have a sound enough background in security and cryptography, so I would hardly be able to make a valuable input yet. Sorry

My best regards to you

**Comment by conradkleinespel** [comment]
@sumproxy: Alright, that's understandable. I'm sure you would be able to contribute in a meaningful way though. Especially since the GUI part probably won't touch the crypto parts that much.

Any specific things you'd like to see for a GUI ? Would you prefer to have some kind of web app ? Or rather a terminal GUI ? Or a native app ?

**Comment by sumproxy** [comment]
I’d personally be happy with a terminal GUI, but I’m not sure how that would be possible. Do you mean some background process or rather running the app in a separate terminal window?

**Comment by conradkleinespel** [comment]
@sumproxy A terminal GUI would be awesome, IMO ! This could be done with the help of [ncurses](https://crates.io/crates/ncurses/). It's a library that makes writing terminal GUIs easier.

Eventually, I'm thinking it would be nice to have a background process that exposes an HTTP API. This API would give access to the Rooster file's data (after authentication of course). This could come in handy when Rooster gets a web client, browser extensions, or mobile apps.

**Comment by sumproxy** [comment]
A web client+a browser extension is definitely superior, but that looks something like a LastPass to me and will probably take a lot of time to write. Let me know how I can be of help

**Comment by conradkleinespel** [comment]
@sumproxy It really depends on what you'd prefer to work on (assuming you want to work on a feature). I could help with one or the other. I myself would probably find the terminal UI more useful, as I spend most of my time in the terminal anyway.

**Comment by sumproxy** [comment]
@conradkleinespel Sorry to keep you waiting. As I found out on OS X ncurses-rs requires installing ncurses with homebrew, which I'm afraid might break someting in my system, so I proceeded with natives. Eventually `cargo build` completed successfully, well at least without errors. But! examples were not built, which leaves me without a place to start.

**Comment by conradkleinespel** [comment]
@sumproxy Oh, right. Ncurses can be somewhat to start using. I have found [this ncurses tutorial](http://tldp.org/HOWTO/NCURSES-Programming-HOWTO/init.html) to be of great help. Also, the `ncurses-rs` package has some useful [examples](https://github.com/jeaye/ncurses-rs/tree/master/examples).

I think you can build an example program by using:
```sh
cargo build --example <example-name>
```

**Comment by sumproxy** [comment]
I had to go with the default recipe for Mac OS and
> force Homebrew to link the library to `/usr/local/lib`: `brew link --force nurses` and set that path to `LIBRARY_PATH` environmental variable.

after that the example compiled. However if we mean to make the installation painless on both Linux and Mac OS I think we should consider this nuisance.
I’ll write back once I have something to report after working with examples.

**Comment by conradkleinespel** [comment]
Definitely, we should document this process or automate it.

Thanks for all the time you're spending on this. I truly appreciate it ! :smiley:

**Comment by conradkleinespel** [comment]
Definitely, we should document this process or automate it.

Thanks for all the time you're spending on this. I truly appreciate it ! :smiley:

**Comment by qmx** [comment]
can't we piggyback on something like gpg-agent or the system's keyring?

**Comment by conradkleinespel** [comment]
@qmx: That would be cool ! I don't think this excludes having a GUI though, does it ? It would be cool to have a GUI regardless. At the moment, it's a bit cumbersome to figure out what app name you have given for a specific password. I tend to revert to `rooster list | grep -i dropbox`, for instance. But that could be vastly improved.

Ideally, we'd have a GUI that you can open without having to retype your password (by going through gpg-agent or the system's keyring, as you say).

Maybe we should open another issue to discuss gpg-agent or system keyring integration so we can keep this GUI issue on point ?

**Comment by conradkleinespel** [comment]
Regarding the GUI, I was just thinking, having something kind of like [dmenu](https://www.youtube.com/watch?v=Iw6sbQ5NLjM) would be amazing.

The UX would work as follows: type Alt-P, a menu pops up with a list of apps you have stored in Rooster. Start typing the app you're looking for. Hit enter. The password is in the clipboard.

For OSX users, you could think of this as the "Spotlight for Passwords".

**Comment by sumproxy** [comment]
I'm really perplexed with analytics tracking in an application of this kind, even if it's a just stub. I can't accept *any* tracking in a password manager.

**Comment by conradkleinespel** [comment]
@sumproxy So as to keep this issue on point, that is "Retyping master password & GUI", I've opened #7. Feel free to share your thoughts on the subject there.

**Comment by conradkleinespel** [comment]
@sumproxy I think suggesting adding tracking was a **mistake on my part**. Proof is it made you go away and I didn't even have time to implement any tracking. So I'm removing this suggestion. Let me know if you are, by any chance, still interested in giving a tracking-free Rooster a chance :smiley: 

I've just removed any reference to tracking from the README and I've removed the stub in the code itself. And I've published a new Rooster version `2.4.0` on Crates.io.
